pyscriptjs/src/python/pyscript/__init__.py

```diff
@@ -709,6 +714,88 @@ def deprecate(name, obj, instead):
     ns["PyScript"] = DeprecatedGlobal("PyScript", PyScript, message)
 
 
+class _PyscriptWebLoop(WebLoop):
+    def __init__(self):
+        orig_loop = asyncio.get_event_loop()
+        super().__init__()
+        asyncio._set_running_loop(orig_loop)
+        self.ready = False
+        self.deferred_handles = []
+
+    def call_later(  # type: ignore[override]
+        self,
+        delay: float,
+        callback: Callable[..., Any],
+        *args: Any,
+        context: contextvars.Context | None = None,
+    ) -> asyncio.Handle:
+        """Arrange for a callback to be called at a given time.
+        Return a Handle: an opaque object with a cancel() method that
+        can be used to cancel the call.
+        The delay can be an int or float, expressed in seconds.  It is
+        always relative to the current time.
+        Each callback will be called exactly once.  If two callbacks
+        are scheduled for exactly the same time, it undefined which
+        will be called first.
+        Any positional arguments after the callback will be passed to
+        the callback when it is called.
+        This uses `setTimeout(callback, delay)`
+        """
+        if delay < 0:
+            raise ValueError("Can't schedule in the past")
+        h = asyncio.Handle(callback, args, self, context=context)
+
+        def run_handle():
+            if h.cancelled():
+                return
+            try:
+                h._run()
+            except SystemExit as e:
+                if self._system_exit_handler:
+                    self._system_exit_handler(e.code)
+                else:
+                    raise
+            except KeyboardInterrupt:
+                if self._keyboard_interrupt_handler:
+                    self._keyboard_interrupt_handler()
+                else:
+                    raise
+
+        if self.ready:
+            setTimeout(create_once_callable(run_handle), delay * 1000)
+        else:
+            self.deferred_handles.append((run_handle, self.time() + delay))
+        return h
+
+    def start_(self):
+        asyncio._set_running_loop(self)
+        t = self.time()
+        for [run_handle, delay] in self.deferred_handles:
+            delay = delay - t
+            if delay < 0:
+                delay = 0
+            setTimeout(create_once_callable(run_handle), delay * 1000)
+        self.ready = True
+        self.deferred_handles = []
+
+
+_LOOP = _PyscriptWebLoop()
+
+
+def _start_loop():
+    _LOOP.start_()
+
+
+@contextmanager
+def _pyscript_event_loop():
+    orig_loop = asyncio.get_event_loop()
+    asyncio._set_running_loop(_LOOP)
```

Comment:
Out of curiousity, if we setting the running loop with our own, is it possible that it may break tasks running on a different loop?

Reply:
Truthfully I'm not fully sure, event loops are complicated. Python was written with the concept that there is just one event loop per process. In this case I think it is okay because:
1. the two event loops are almost the same and have the same futures and tasks, so if the wrong loop is invoked it still works correctly
2. the two event loops have trivial scheduling logic (since they are just using the browser event loop)

I think there is a better approach which would be rather than switching between two event loops, use a flag to indicate whether we're in a user-context where tasks should be deferred or in a host-context where tasks should not be deferred.

Reply:
Okay I applied this new approach.